src/dsp_tools/commands/validate_data/sparql/value_shacl.py

```diff
@@ -52,6 +53,61 @@ def _add_property_shapes_to_class_shapes(onto: Graph) -> Graph:
     return Graph()
 
 
+def _construct_unique_value_shape(onto: Graph) -> Graph:
+    query_s = '''
+    PREFIX owl: <http://www.w3.org/2002/07/owl#> 
+    PREFIX sh: <http://www.w3.org/ns/shacl#>
+    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
+    PREFIX salsah-gui: <http://api.knora.org/ontology/salsah-gui/v2#> 
+    PREFIX api-shapes: <http://api.knora.org/ontology/knora-api/shapes/v2#>
+    PREFIX knora-api:  <http://api.knora.org/ontology/knora-api/v2#>
+
+    CONSTRUCT {
+
+    ?uniqueShapeIRI
+      a              sh:NodeShape ;
+      sh:targetClass ?class ;
+      sh:sparql   [
+            a          sh:SPARQLConstraint ;
+            sh:message "A resource may not have the same property and value more than one time." ;
+            sh:select  """
+            PREFIX rdfs:       <http://www.w3.org/2000/01/rdf-schema#>
+            PREFIX knora-api:  <http://api.knora.org/ontology/knora-api/v2#>
+            PREFIX api-shapes: <http://api.knora.org/ontology/knora-api/shapes/v2#>
+
+                SELECT $this ?path ?value WHERE {
+
+                    $this ?path ?valueClass .
+                    
+                    {
+                        ?prop rdfs:subPropertyOf knora-api:valueHas .
+                        ?valueClass ?prop ?value .
+                    }
+                    UNION
+                    {
+                        ?valueClass knora-api:valueAsString|api-shapes:linkValueHasTargetID|api-shapes:listNodeAsString ?value .
```

Comment:
I very much don't like this query. I have tried different approaches without success.
But I think this is an optimisation we can take a look at in the future.